Update `FunctionalAnnotationAggMember` class for compatibility with MetaP Aggregation tables and implement migrator

kheal opened the pull request:
This was moved from the berkeley fork to this repo.  Original PR found here: https://github.com/microbiomedata/berkeley-schema-fy24/pull/242

# Description
This PR will 
1) Modify the existing `FunctionalAnnotationAggMember` class to be able to accomodate both annotations from metagenomics and metaproteomics analyses. This is accomplished by exchanging the slot `metagenome_anlaysis_id` with the existing slot `was_generated_by`.
2) Implement a migrator to adjust existing records.  
3) Deprecate the `metagenome_anlaysis_id` slot as there are now no classes using this slot.
5) Update tests accordingly
6) Add description to `count` slot
7) ~~Add a unique_keys to `FunctionalAnnotationAggMember` class that is a combination of `was_generated_by`and`gene_function_id`~~ See #2155 for why this was removed from this PR.

This PR will close [1253](https://github.com/microbiomedata/nmdc-schema/issues/1253)

# PR Information

## What type of PR is this? (check all applicable)
- [x] Schema change: Structure and content
  - deleted a `slot`
     
## Related Issues

- Fixes: [1253](https://github.com/microbiomedata/nmdc-schema/issues/1253)

## Did you add/update any tests?
- [X] Yes
    - All example yamls are updated (including invalid examples
    - Doctest for migrator added
    - Added invalid example for FunctionalAnnotationAggMember to check for rule on count slot

## Could this schema change make it so any valid data becomes invalid?
- [x] Yes _(A migrator is required)_

## If you answered "Yes", does this PR branch include that migrator?
- [x] Yes

## Does this PR have any downstream implications?
- [x] Yes _(Explain below)_
_We will need to update the aggregation scripts (issue filed here: https://github.com/microbiomedata/nmdc-aggregator/issues/13) and the data portal to use the new slot_

eecavanna commented on nmdc_schema/migrators/migrator_from_XX_to_PR242.py:
```diff
@@ -0,0 +1,32 @@
+        {'id': 123, 'count': 1, 'was_generated_by': 'mgm123'}
+        
+        """
+        self.logger.info(f"Migrating Extraction: {fun_agg['id']}")
```
I was surprised to see the entity being referred to as an "extraction" here. Maybe that is the result of a copy/paste/edit sequence, where the "edit" step hasn't taken place yet (given that this PR is in a draft state).

eecavanna commented on nmdc_schema/migrators/migrator_from_XX_to_PR242.py:
```diff
@@ -0,0 +1,32 @@
+        r"""Migrates the database from conforming to the original schema, to conforming to the new schema."""
+
+        self.adapter.process_each_document(
+            "functional_annotation_agg", [self.move_metagenome_id_to_was_generated_by]
```
Note to self: This collection doesn't have an `id` slot. Ensure the migration framework (both the adapter methods and the standard code in the downstream migration notebooks) supports this.

kheal replied:
I removed the id field from the doctests as well.  This PR should enable _compound indexing_ in mongoDB on this collection (the unique_keys aspect of the PR).  I'm not sure that's related here but I wanted to keep that in the conversation since I know this collection is a bit unconventional.

kheal replied:
Brynn showed me how to test the migrator locally and it passed!

eecavanna replied:
Great! Thanks for confirming. Similarly, I confirmed the adapter methods and migration notebook don't depend on there being a field named `id` in the document (whew).

shreddd commented:
Prior to merging / deploying this, I wanted to make sure that the needed changes also get rolled into 
https://github.com/microbiomedata/nmdc-aggregator/
Otherwise this will break when a new collection is generated from the aggregator. 

Possibly related? https://github.com/microbiomedata/nmdc-aggregator/issues/24

kheal commented:
> Prior to merging / deploying this, I wanted to make sure that the needed changes also get rolled into https://github.com/microbiomedata/nmdc-aggregator/ Otherwise this will break when a new collection is generated from the aggregator.
> 
> Possibly related? [microbiomedata/nmdc-aggregator#24](https://github.com/microbiomedata/nmdc-aggregator/issues/24)

@shreddd - Yes that is the companion issue to keep the aggregator in step with this change.

kheal commented:
@sierra-moxon and @turbomam Could you please check that I've implemented the use of `unique_keys` correctly?

aclum commented:
This PR looks good. However, I don't see a linked issue in nmdc-server to remove ingesting metap_gene_function_aggregation so we should make sure @naglepuff @marySalvi have time to do that before Nov 11 (for dev testing, Nov monthly release is scheduled for Nov 18) before this gets merged in.

aclum commented:
Created companion ticket https://github.com/microbiomedata/nmdc-server/issues/1431

kheal commented:
@eecavanna or @brynnz22 - Do either of you have guidance for how the migrator should be named?  It is currently `nmdc_schema/migrators/migrator_from_XX_to_PR2203.py`.  I can either leave it as that or update, but otherwise I think this MR is about ready to be merged in.

brynnz22 commented:
Maybe `migrator_from_11_0_3_to_11_1_0`. I'm not sure what the next version of the schema will be. @turbomam do you have an ideas what the next version of the schema will be after merging this in?

eecavanna commented:
Sorry about the delay! In a "workshop" all day today and Wednesday.

I agree with @brynnz22's recommendation. The "from" version is the currently released schema version. The "to" version is still TBD. The value @brynnz22 suggested makes sense to me as a placeholder (sufficient for merging, in my opinion, given that we don't know what the next schema version will be).

---

The requirement of putting the "to" version in the migrator puts us in this chicken-and-egg situation. I do plan to change this part of the procedure "soon" ™.

brynnz22 commented:
@kheal @eecavanna Since I am working on a migrator right now for the `has_calibration` slot, it sounds like this should be a partial migrator and same with the one I am working on. Because they will both probably feed into the same schema update. Does this seem right? My PR will be ready sometime today.

kheal commented:
@brynnz22 that seems reasonable.  #2204 also has a migrator that could be bundled into this if we decide to go down that route.

kheal commented:
@turbomam - do you want to review this before we merge it in?

turbomam commented on src/schema/nmdc.yaml:
```diff
@@ -1043,19 +1050,21 @@ slots:
+
   count:
     range: integer
+    description: The number of sequences (for a metagenome or metatranscriptome) or spectra (for metaproteomics) associated with the specified function.
```
<3

turbomam replied:
I would like to move that definition for count to `slot_usage`. The `count` slot could easily be used with different meanings in other class contexts.

turbomam commented on src/schema/nmdc.yaml:
```diff
@@ -1043,19 +1050,21 @@ slots:
     range: WorkflowExecution
     description: The identifier for the analysis activity that generated the functional annotation results, where the analysis activity is an instance of the/an appropriate subclass of WorkflowExecution
     required: true
-    any_of:
```
<3

turbomam commented on src/schema/nmdc.yaml:
```diff
@@ -299,16 +299,23 @@ classes:
-      metagenome_annotation_id:
-        structured_pattern: # doesn't include act
-          syntax: "{id_nmdc_prefix}:(wfmgan|wfmtan)-{id_shoulder}-{id_blade}{id_version}$"
+      was_generated_by:
```
@aclum and others: are we simplifying the range of [was_generated_by](https://microbiomedata.github.io/nmdc-schema/was_generated_by/) at some point?

kheal replied:
See https://github.com/microbiomedata/nmdc-schema/pull/2246

turbomam commented:
> @sierra-moxon and @turbomam Could you please check that I've implemented the use of `unique_keys` correctly?

@kheal If you have tests that show valid and invalid re-use of either of the identifying slots, then I approve for merging.

turbomam reviewed (commented):
This will be a great PR with a few tweaks. Let me know if you think any of my requests will be difficult (or harmful!)

kheal commented:
@turbomam.  Thanks for the input.  

@turbomam  or @sierra-moxon I could use some help with the test and/or the `unique_keys` slot. I added a test in the problem/invalid folder - it is passing validation but it shouldn't, so something is wrong, but I can't tell if it's the testing or the actual definition in the schema. I used the [`Isotope` example here ](https://linkml.io/linkml/schemas/constraints.html) as guidance for how to use the `unique_keys` slot.

I've moved the description for the `count` slot to the class's `slot_usage`.

kheal commented:
After consulting with @sierra-moxon, we've determined that uniqueness across objects is not supported in JSONSchema, because linkml-validate uses the jsonschema serialization of the model to do validation, it is not checking the unique keys constraint.  See #2155 for more discussion

Because of the lack of verifiability on the `unique_keys` constraint currently supported, I've removed that from this PR.  Thank you @turbomam for encouraging adding a testing file for this where we uncovered this.
